File: lua/perfect-numbers/perfect-numbers.lua
local function aliquot_sum(n)
  local sum = 0

  for i = 1, n / 2 do
    if n % i == 0 then
      sum = sum + i
    end
  end

  return sum
end

local function classify(n)
  local sum = aliquot_sum(n)

  if sum < n then
    return "deficient"
  elseif sum > n then
    return "abundant"
  else
    return "perfect"
  end
end

return {
  aliquot_sum = aliquot_sum,
  classify = classify
}

File: csharp/perfect-numbers/PerfectNumbers.cs
using System;
using System.Linq;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number <= 0)
            throw new ArgumentOutOfRangeException();

        var sum = AliquotSum(number);

        if (sum < number)
            return Classification.Deficient;
        else if (sum > number)
            return Classification.Abundant;
        else
            return Classification.Perfect;
    }

    private static int AliquotSum(int number)
    {
        return Enumerable.Range(1, number / 2)
            .Where(i => number % i == 0)
            .Sum();
    }
}
